### PR TITLE
Feature/vehnt

### PR DESCRIPTION
**Pull request by ThornM9**

### CODE REVIEW

File: packages/helium-sub-daos-sdk/src/pdas.ts [hunk 1]
```diff
@@ -35,3 +35,14 @@ export function subDaoKey(
...
   );
 }
+
+export function stakePositionKey(
+  voterAuthority: PublicKey,
+  depositEntryIdx: number,
+  programId: PublicKey = PROGRAM_ID
+): [PublicKey, number] {
+  return PublicKey.findProgramAddressSync([
+    Buffer.from("stake_position", "utf-8"), voterAuthority.toBuffer(), (new BN(depositEntryIdx)).toBuffer()],
```
Comment (ChewingGlass) [inline packages/helium-sub-daos-sdk/src/pdas.ts:45]: Not sure I love the new BN approach here. Shouldn't this be an explicit number of bytes, and the endianess specified?

File: programs/helium-sub-daos/src/instructions/initialize_sub_dao_v0.rs [hunk 2]
```diff
@@ -102,7 +105,26 @@ pub struct InitializeSubDaoV0<'info> {
...
+  pub rewards_escrow: Box<Account<'info, TokenAccount>>, // TODO why can this just be any tokenaccount?
+
+  /// CHECK: Initialized via cpi
+  #[account(
+    mut,
+    seeds = ["account_windowed_breaker".as_bytes(), staker_pool.key().as_ref()],
+    seeds::program = circuit_breaker_program.key(),
+    bump
+  )]
+  pub staker_pool_circuit_breaker: AccountInfo<'info>,
```
Comment (ChewingGlass) [inline programs/helium-sub-daos/src/instructions/initialize_sub_dao_v0.rs:117]: hurrah for circuit breaker

File: programs/helium-sub-daos/src/instructions/issue_rewards_v0.rs [hunk 3]
```diff
@@ -107,57 +152,59 @@ pub fn handler(ctx: Context<IssueRewardsV0>, args: IssueRewardsArgsV0) -> Result
...
+    ctx.accounts.mint_dnt_emissions_ctx().with_signer(&[&[
+      b"sub_dao",
+      ctx.accounts.dnt_mint.key().as_ref(),
+      &[ctx.accounts.sub_dao.bump_seed],
+    ]]),
+    MintArgsV0 {
+      amount: total_emissions
+        .checked_div(100.try_into().unwrap())
+        .unwrap()
+        .checked_mul(94)
```
Comment (ChewingGlass) [inline programs/helium-sub-daos/src/instructions/issue_rewards_v0.rs:172]: Ordering matters here, right? Don't we lose precision by dividing by 100 first? Should do it after the mul.

File: programs/helium-sub-daos/src/instructions/staking/claim_rewards_v0.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  #[account(
+    mut,
+    has_one = staker_pool,
+    has_one = dnt_mint,
+  )]
+  pub sub_dao: Account<'info, SubDaoV0>,
+  pub dnt_mint: Box<Account<'info, Mint>>,
+
+  #[account(
+    seeds = ["sub_dao_epoch_info".as_bytes(), sub_dao.key().as_ref(), &args.epoch.to_le_bytes()], // Break into 30m epochs
```
Comment (ChewingGlass) [inline programs/helium-sub-daos/src/instructions/staking/claim_rewards_v0.rs:54]: Probably have this elsewhere but epoch is 24h not 30m

**Review by ChewingGlass — CHANGES_REQUESTED** [review]

File: packages/helium-sub-daos-sdk/src/pdas.ts (hunk 1 above)
Comment (ThornM9) [reply packages/helium-sub-daos-sdk/src/pdas.ts:45]: The index is only 1 byte, so endianess doesn't matter. I'll take a look and see if there's a better way to write it

File: programs/helium-sub-daos/src/instructions/staking/claim_rewards_v0.rs (hunk 4 above)
Comment (ThornM9) [reply programs/helium-sub-daos/src/instructions/staking/claim_rewards_v0.rs:54]: Removed all instances of this in my next pr

File: programs/helium-sub-daos/src/instructions/issue_rewards_v0.rs (hunk 3 above)
Comment (ThornM9) [reply programs/helium-sub-daos/src/instructions/issue_rewards_v0.rs:172]: Yep definitely, fixed

File: packages/helium-sub-daos-sdk/src/pdas.ts (hunk 1 above)
Comment (ChewingGlass) [reply packages/helium-sub-daos-sdk/src/pdas.ts:45]: const b = Buffer.alloc(1)
b.writeSomethingLE(number)

b

File: programs/helium-sub-daos/src/instructions/staking/stake_v0.rs [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+      || sd_acc_info.key() == Pubkey::default()
+    {
+      continue;
+    }
+
+    assert!(
+      stake_position.last_claimed_epoch == 0
+        || stake_position.allocations[i].sub_dao == sd_acc_info.key()
+    ); // must be new account or the subdao keys should be equal
+    assert!(sd_acc_info.is_writable);
```
Comment (ChewingGlass) [inline programs/helium-sub-daos/src/instructions/staking/stake_v0.rs:140]: It seems like we're missing a check that the sub_dao is part of the right dao with the right hnt?

**Review by ChewingGlass — CHANGES_REQUESTED** [review]

File: programs/helium-sub-daos/src/instructions/staking/close_stake_v0.rs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  // don't allow unstake without claiming available rewards
+  let curr_epoch = current_epoch(ctx.accounts.clock.unix_timestamp);
+  assert!(ctx.accounts.stake_position.last_claimed_epoch >= curr_epoch - 1);
+
+  // position_vehnt = available_vehnt * hnt_amount / amount_deposited_native
+  // let position_vehnt = available_vehnt
+  //   .checked_mul(ctx.accounts.stake_position.hnt_amount)
+  //   .unwrap()
+  //   .checked_div(d_entry.amount_deposited_native)
+  //   .unwrap();
```
[the remark this answers is not in the document]
Comment (ThornM9) [reply programs/helium-sub-daos/src/instructions/staking/close_stake_v0.rs:67]: Old code, deleted

File: programs/helium-sub-daos/src/instructions/staking/purge_position_v0.rs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      continue;
+    }
+    assert!(stake_position.allocations[i].sub_dao == sd_acc_info.key());
+    assert!(sd_acc_info.is_writable);
+
+    let mut sub_dao_data = sd_acc_info.try_borrow_mut_data()?;
+    let mut sub_dao_data_slice: &[u8] = &sub_dao_data;
+    let sub_dao = &mut SubDaoV0::try_deserialize(&mut sub_dao_data_slice)?;
+
+    update_subdao_vehnt(sub_dao, curr_ts);
```
[the remark this answers is not in the document]
Comment (ThornM9) [reply programs/helium-sub-daos/src/instructions/staking/purge_position_v0.rs:109]: I don't think it makes that assumption, the endpoint updates the subdao_vehnt to the curr_ts, then uses the time_since_expiry to apply the correction factor. Correct me if I'm wrong or misunderstanding

File: programs/helium-sub-daos/src/instructions/staking/refresh_position_v0.rs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  let voter = ctx.accounts.vsr_voter.load()?;
+  let registrar = &ctx.accounts.registrar.load()?;
+  let d_entry = voter.deposits[args.deposit_entry_idx as usize];
+  let voting_mint_config = &registrar.voting_mints[d_entry.voting_mint_config_idx as usize];
+  let curr_ts = registrar.clock_unix_timestamp();
+  let available_vehnt = d_entry.voting_power(voting_mint_config, curr_ts)?;
+  let future_vehnt = d_entry.voting_power(voting_mint_config, curr_ts + 1)?;
+  let fall_rate = available_vehnt.checked_sub(future_vehnt).unwrap();
+
+  if ctx.accounts.stake_position.hnt_amount <= d_entry.amount_deposited_native {
```
[the remark this answers is not in the document]
Comment (ThornM9) [reply programs/helium-sub-daos/src/instructions/staking/refresh_position_v0.rs:52]: Fixed

File: programs/helium-sub-daos/src/instructions/staking/refresh_position_v0.rs [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    sub_dao.vehnt_staked = sub_dao.vehnt_staked.checked_sub(vehnt_diff).unwrap();
+    sub_dao.vehnt_fall_rate = sub_dao.vehnt_fall_rate.checked_sub(fall_rate_diff).unwrap();
+
+    sub_dao.try_serialize(&mut *sub_dao_data)?;
+  }
+
+  // update the stake position
+  stake_position.fall_rate = fall_rate;
+  stake_position.hnt_amount = d_entry.amount_deposited_native;
+  stake_position.last_claimed_epoch = current_epoch(curr_ts);
```
[the remark this answers is not in the document]
Comment (ThornM9) [reply programs/helium-sub-daos/src/instructions/staking/refresh_position_v0.rs:112]: Addressed

File: programs/helium-sub-daos/src/instructions/staking/stake_v0.rs (hunk 5 above)
Comment (ThornM9) [reply programs/helium-sub-daos/src/instructions/staking/stake_v0.rs:140]: Yep, added that check to all the relevant instructions. Also added a missing check on registrar seeds

**Review by ChewingGlass — APPROVED** [review]

File: programs/helium-sub-daos/src/utils.rs [hunk 10]
```diff
@@ -28,3 +30,96 @@ pub fn current_epoch(unix_timestamp: i64) -> u64 {
...
+  require_gte!(
+    max_locked_vote_weight,
+    locked_vote_weight,
+    ErrorCode::FailedVotingPowerCalculation,
+  );
+
+  baseline_vote_weight
+    .checked_add(locked_vote_weight)
+    .ok_or_else(|| error!(ErrorCode::FailedVotingPowerCalculation))
+}
```
Comment (bryzettler) [inline programs/helium-sub-daos/src/utils.rs:95]: In the rewrite we have these, and `voting_mint_config.locked_vote_weight`
```rust
    locked_vote_weight_scaled_factor: u64,
    minimum_required_lockup_secs: u64,
    max_extra_lockup_vote_weight_scaled_factor: u64,
    genesis_vote_power_multiplier: u8,
    genesis_vote_power_multiplier_expiration_ts: i64,
```
Im going to be moving the changes over to this repo so take a look at those changes too

**Pull request merged**